#! /usr/bin/env python3
### vmshed: { "vms_all": [3], "vms_ci": [3], "drbd_version_min": "9.0" }
#
# Interesting that this issue survived for so long!!
# Consider 3 nodes: A, B, C. (A = Primary, Quorum enabled)
# Down C (gracefully)
# Down B (gracefully, allows A to keep quorum due to "Last man standing"
# Let C resync to A
# Connect B to C and see a split brain up to DRBD 9.0.21!!

# TODO: Execute all possible orders of nodes joining the primary

from python import drbdtest
from python.drbdtest import connections, log, peer_devices
from subprocess import CalledProcessError

resource = drbdtest.setup(nodes=3)
resource.resource_options = 'quorum majority; on-no-quorum io-error;'
resource.disk_options = 'c-max-rate 250k;'
resource.add_disk('10M')

resource.up_wait()

log('* Make up-to-date data available.')
resource.skip_initial_sync()

[node_a, node_b, node_c] = resource.nodes
dev_name = node_a.volumes[0].device()

node_a.primary()
connections_to_a = connections(to_node=node_a)
connections_to_a.event(r'connection .* role:Primary')

connections(node_c).disconnect()
node_a.write()

connections(node_b).disconnect()
node_a.write()

connections(node_c, node_a).connect()
node_c.volumes.event(r'device .* disk:UpToDate')

connections_bc = drbdtest.Connections()
connections_bc.bidir_add(node_b, node_c)
connections_bc.connect()
connections_bc.event(r'connection .* connection:Connected')

connection_bc = connections(node_b, node_c)
peer_devices_bc = drbdtest.PeerDevices.from_connections(connection_bc)
peer_devices_bc.event(r'peer-device .* replication:SyncTarget')

connection_ba = connections(node_b, node_a)
connection_ba.connect();
connection_ba.event(r'connection .* connection:Connected')

node_b.volumes.event(r'device .* disk:UpToDate')

log('* Shut down and clean up.')
resource.down()
resource.teardown()
